pallets/connectors-gateway/axelar-gateway-precompile/src/lib.rs

```diff
@@ -0,0 +1,397 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use cfg_types::domain_address::{Domain, DomainAddress};
+use codec::alloc::string::ToString;
+use ethabi::Token;
+use fp_evm::PrecompileHandle;
+use pallet_evm::{ExitError, PrecompileFailure};
+use precompile_utils::prelude::*;
+use sp_core::{bounded::BoundedVec, ConstU32, H160, H256, U256};
+use sp_runtime::DispatchResult;
+use sp_std::vec::Vec;
+
+pub const MAX_SOURCE_CHAIN_BYTES: u32 = 128;
+pub const MAX_SOURCE_ADDRESS_BYTES: u32 = 32;
+pub const MAX_TOKEN_SYMBOL_BYTES: u32 = 32;
+pub const MAX_PAYLOAD_BYTES: u32 = 1024;
+pub const PREFIX_CONTRACT_CALL_APPROVED: [u8; 32] = keccak256!("contract-call-approved");
+
+pub type String<const U32: u32> = BoundedString<ConstU32<U32>>;
+pub type Bytes<const U32: u32> = BoundedBytes<ConstU32<U32>>;
+
+pub use pallet::*;
+
+#[derive(
+	PartialEq,
+	Clone,
+	codec::Encode,
+	codec::Decode,
+	scale_info::TypeInfo,
+	codec::MaxEncodedLen,
+	frame_support::RuntimeDebugNoBound,
+)]
+pub struct SourceConverter {
+	domain: Domain,
+}
+
+impl SourceConverter {
+	pub fn try_convert(&self, maybe_address: &[u8]) -> Option<DomainAddress> {
+		match self.domain {
+			Domain::Centrifuge => Some(DomainAddress::Centrifuge(Self::try_into_32bytes(
+				maybe_address,
+			)?)),
+			Domain::EVM(id) => Some(DomainAddress::EVM(
+				id,
+				Self::try_into_20bytes(maybe_address)?,
+			)),
+		}
+	}
+
+	fn try_into_32bytes(maybe_address: &[u8]) -> Option<[u8; 32]> {
+		if maybe_address.len() == 32 {
+			let mut address: [u8; 32] = [0u8; 32];
+			address.copy_from_slice(maybe_address);
+
+			Some(address)
+		} else {
+			None
+		}
+	}
+
+	fn try_into_20bytes(maybe_address: &[u8]) -> Option<[u8; 20]> {
+		if maybe_address.len() == 20 {
+			let mut address: [u8; 20] = [0u8; 20];
+			address.copy_from_slice(maybe_address);
+
+			Some(address)
+		} else {
+			None
+		}
+	}
+}
```

Comment:
We could use `vec_to_fixed_array` as the implementation logic of these functions aside from the `len()` check

Reply:
ah I knew you had this problem already once. I will keep it as is fro now, as I need the len check.